lib/wasi/src/state.rs

```diff
@@ -20,75 +20,297 @@ use wasmer_runtime_core::debug;
+    /// Expected a file but found not a file
+    NotAFile,
+    /// The fd given was not usable
+    InvalidFd,
+    /// File exists
+    AlreadyExists,
+    /// Something failed when doing IO. These errors can generally not be handled.
+    /// It may work if tried again.
+    IOError,
+    /// A WASI error without an external name.  If you encounter this it means
+    /// that there's probably a bug on our side (maybe as simple as forgetting to wrap
+    /// this error, but perhaps something broke)
+    UnknownError(__wasi_errno_t),
+}
+
+impl WasiFsError {
+    pub fn from_wasi_err(err: __wasi_errno_t) -> WasiFsError {
+        match err {
+            __WASI_EBADF => WasiFsError::InvalidFd,
+            __WASI_EEXIST => WasiFsError::AlreadyExists,
+            __WASI_EIO => WasiFsError::IOError,
+            _ => WasiFsError::UnknownError(err),
+        }
+    }
 }
 
-impl WasiFile {
-    pub fn close(self) {}
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+/// it does not require `Drop` because `std::fs::File` does not implement it
+/// (TODO: figure out how that makes sense, because it clearly does)
+pub trait WasiFile: std::fmt::Debug + Write + Read + Seek {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64;
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
 }
 
-impl Write for WasiFile {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match self {
-            WasiFile::HostFile(hf) => hf.write(buf),
-        }
+impl WasiFile for fs::File {
+    fn last_accessed(&self) -> u64 {
+        self.metadata()
+            .unwrap()
+            .accessed()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
     }
 
-    fn flush(&mut self) -> io::Result<()> {
-        match self {
-            WasiFile::HostFile(hf) => hf.flush(),
-        }
+    fn last_modified(&self) -> u64 {
+        self.metadata()
+            .unwrap()
+            .modified()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn created_time(&self) -> u64 {
+        self.metadata()
+            .unwrap()
+            .created()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
     }
 
+    fn size(&self) -> u64 {
+        self.metadata().unwrap().len()
+    }
+}
+
+#[derive(Debug)]
+pub struct Stdout(std::io::Stdout);
+impl Read for Stdout {
```

Comment:
I feel like a 'nicer' way of doing this is to essentially encapsulate Read/Write/Seek in our trait, and provide default generic implementations if the type implements Read/Write/Seek.

(this is probably *not* nicer in many other ways, so meh.)

Reply:
Hmm, yeah, I see what you mean.  I think this is fine because while it is garbage code, we only need to write it once -- and I'll refactor the `state.rs` file to hide things like this because we mostly never need to see it -- and it lets us use all code working with the standard traits including third party libraries which could save us a lot of time in the future.

Reply:
I mean, so does this approach. The issue (either way, really) is that if a third party library is missing any one of these three, you have to wrap it in a newtype and create dummy implementations. At least with the 'wrapper trait' approach, you aren't implementing, say, Seek on something that can't actually support Seek. *shrug*

only solution I can think of that doesn't require boilerplate is probably 'wrapper trait' + specialization, and even then a) i don't think we want to do that and b) it doesn't appear to work like it does in my head so...